bitcoin/src/pow.rs

```diff
@@ -0,0 +1,1530 @@
+
+impl Div for U256 {
+    type Output = Self;
+    fn div(self, rhs: Self) -> Self { self.div_rem(rhs).0 }
+}
+
+impl Rem for U256 {
+    type Output = Self;
+    fn rem(self, rhs: Self) -> Self { self.div_rem(rhs).1 }
+}
+
+impl Not for U256 {
+    type Output = Self;
+
+    fn not(self) -> Self { U256(!self.0, !self.1) }
+}
+
+impl Shl<u32> for U256 {
+    type Output = Self;
+    fn shl(self, shift: u32) -> U256 { self.wrapping_shl(shift) }
+}
+
+impl Shr<u32> for U256 {
+    type Output = Self;
+    fn shr(self, shift: u32) -> U256 { self.wrapping_shr(shift) }
+}
+
+impl fmt::Display for U256 {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.is_zero() {
+            f.pad_integral(true, "", "0")
+        } else {
+            self.fmt_decimal(f)
+        }
+    }
+}
+
+impl fmt::Debug for U256 {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { write!(f, "{:#x}", self) }
+}
+
+macro_rules! impl_hex {
+    ($hex:ident, $fmt:literal) => {
+        impl $hex for U256 {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                if f.alternate() {
+                    write!(f, "0x")?;
+                }
+                write!(f, $fmt, self.0)?;
+                write!(f, $fmt, self.1)?;
+                Ok(())
+            }
+        }
+    };
+}
+impl_hex!(LowerHex, "{:032x}");
+impl_hex!(UpperHex, "{:032X}");
+
+#[cfg(feature = "serde")]
+#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
+impl crate::serde::Serialize for U256 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: crate::serde::Serializer,
+    {
+        use crate::hashes::hex::ToHex;
+        let bytes = self.to_be_bytes();
+        if serializer.is_human_readable() {
+            serializer.serialize_str(&bytes.to_hex())
+        } else {
+            serializer.serialize_bytes(&bytes)
+        }
+    }
+}
+
+#[cfg(feature = "serde")]
+#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
+impl<'de> crate::serde::Deserialize<'de> for U256 {
+    fn deserialize<D: crate::serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        use core::convert::TryInto;
+
+        use crate::hashes::hex::FromHex;
+        use crate::serde::de;
+        struct Visitor;
+        impl<'de> de::Visitor<'de> for Visitor {
+            type Value = U256;
+
+            fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                write!(f, "32 bytes or a hex string with 64 characters")
+            }
+
+            fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                if s.len() != 64 {
+                    return Err(de::Error::invalid_length(s.len(), &self));
+                }
+
+                let v = crate::prelude::Vec::from_hex(s)
+                    .map_err(|_| de::Error::invalid_value(de::Unexpected::Str(s), &self))?;
+
+                let mut b = [0_u8; 32];
```

Comment:
`FromHex<[u8; 32]>` is available, so you could use that instead and remove the vast majority of this function along with an allocation. :)

Reply:
Done!